src/internal/storage/fileset/storage.go

```diff
@@ -240,18 +241,37 @@ func (s *Storage) getPrimitives(ctx context.Context, ids []ID) ([]*Primitive, er
 // The path ranges must be non-overlapping and the ranges must be lexigraphically sorted.
 // Concat always returns the ID of a primitive fileset.
 func (s *Storage) Concat(ctx context.Context, ids []ID, ttl time.Duration) (*ID, error) {
-	var size int64
-	additive := index.NewWriter(ctx, s.ChunkStorage(), "additive-index-writer")
-	deletive := index.NewWriter(ctx, s.ChunkStorage(), "deletive-index-writer")
-	for _, id := range ids {
-		md, err := s.store.Get(ctx, id)
-		if err != nil {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	eg, ctx := errgroup.WithContext(cancelCtx)
+	sem := semaphore.NewWeighted(10)
```

Comment:
The semaphore + errgroup might be simplified with SetLimit.
https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit

Reply:
The semantics aren't ideal with that functionality. We want to exit immediately if the context is canceled, not go through all of the goroutines.